File: src/main/java/com/hileco/cortex/primitives/layer/MapLayer.java
package com.hileco.cortex.primitives.layer;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class MapLayer<K, V> extends Layer<MapLayer<K, V>> {
    private final Map<K, V> entries;
    private final Set<K> deletions;

    public MapLayer(MapLayer<K, V> parent) {
        super(parent);
        this.entries = new HashMap<>();
        this.deletions = new HashSet<>();
    }

    public Map<K, V> getEntries() {
        return entries;
    }

    public Set<K> getDeletions() {
        return deletions;
    }

    @Override
    boolean isEmpty() {
        return entries.isEmpty() && deletions.isEmpty();
    }
}

File: cortex-data/src/main/kotlin/com/hileco/cortex/data/layer/MapLayer.kt
package com.hileco.cortex.data.layer

import java.util.HashMap
import java.util.HashSet

class MapLayer<K, V>(parent: MapLayer<K, V>?) : Layer<MapLayer<K, V>>(parent) {
    val entries: MutableMap<K, V> = HashMap()
    val deletions: MutableSet<K> = HashSet()

    override val isEmpty: Boolean
        get() = entries.isEmpty() && deletions.isEmpty()
}
